instance/tests/test_instance_upgrade.py

```diff
@@ -0,0 +1,252 @@
+
+    @staticmethod
+    def _make_git_spec(url, revision):
+        """
+        Helper method to cosntruct GitVersionSpec
+        """
+        return GitVersionSpec(url, revision)
+
+
+@ddt.ddt
+class TestInstanceUpgrade(BaseInstanceUpgradeTest):
+    """
+    Tests for generic instance upgrade functionality
+    """
+
+    class DummyInstanceUpgrade(InstanceUpgrade):
+        """
+        InstanceUpgrade class is intended to be abstract; this class provides simplest implementation possible
+        """
+        INITIAL_RELEASE = "QWERTY"
+        TARGET_RELEASE = "ASDFGH"
+
+        def upgrade_instance(self, instance):
+            """
+            Upgrade single instance
+            """
+            pass
+
+    def setUp(self):
+        """
+        Test set up method
+        """
+        super(TestInstanceUpgrade, self).setUp()
+        self.upgrader = self.DummyInstanceUpgrade()
+
+    def test_get_instances_to_upgrade(self):
+        """
+        Test get_instances_to_upgrade
+        """
+        expected_instances = ['isntance1', 'instance2', 'instance3']
+        self.instance_objects_mock.filter.return_value = expected_instances
+        actual_instances = self.upgrader.get_instances_to_upgrade()
+
+        self.assertEqual(actual_instances, expected_instances)
+        self.instance_objects_mock.filter.assert_called_once_with(
+            active_appserver___status="running",
+            openedx_release__contains=self.DummyInstanceUpgrade.INITIAL_RELEASE,
+            use_ephemeral_databases=False
+        )
+
+    def test_set_git_specs(self):
+        """
+        Test updating git specs
+        """
+        instance_mock = self._make_instance_mock()
+        self.upgrader.set_git_specs(instance_mock)
+
+        expected_platform_spec = self.DummyInstanceUpgrade.EDX_PLATFORM_SPEC
+        expected_config_spec = self.DummyInstanceUpgrade.CONFIGURATION_SPEC
+
+        self._assert_git_specs(instance_mock, expected_platform_spec, expected_config_spec)
+
+    @ddt.unpack
+    @ddt.data(
+        ("", "", "", ""),  # empty
+        ("Test", "Test", "Hit", "Hit"),  # replacement miss
+        ("Test", "Miss", "Miss", "Test"),  # replacement miss
+        ("ASD\nZXC", "ASD", "QWE", "QWE\nZXC"),  # multiline replacement - single hit
+        ("ASD\nASD", "ASD", "QWE", "QWE\nQWE"),  # multiline replacement - multiple hits
+        ("ASD\nASD", "ASD\n", "", "ASD"),  # remove line
+    )
+    def test_replace_extra_config(self, initial_settings, target, replacement, expected_settings):
+        """
+        Test greping configuration_extra_settings
+        """
+        instance_mock = self._make_instance_mock()
+        instance_mock.configuration_extra_settings = initial_settings
+        self.upgrader.replace_extra_config(instance_mock, target, replacement)
+        self.assertEqual(instance_mock.configuration_extra_settings, expected_settings)
+
+    def test_upgrade_instance(self):
```

Comment:
@e-kolpakov Probably just a typo, but this should be called `test_upgrade_instances` :)

When you make this change, you could also update the docstring to something like "Test upgrading multiple instances", to match the pattern you established for the `test_upgrade_instance` methods below.